File: packages/bot/scripts/register-webhook.ts
#!/usr/bin/env node
// Run after deploy: pnpm --filter @nudge/bot register-webhook
// Env vars required: TELEGRAM_BOT_TOKEN, TELEGRAM_WEBHOOK_SECRET, APP_URL

const token = process.env["TELEGRAM_BOT_TOKEN"];
const secret = process.env["TELEGRAM_WEBHOOK_SECRET"];
const appUrl = process.env["APP_URL"];

if (!token || !secret || !appUrl) {
  console.error(
    "Required: TELEGRAM_BOT_TOKEN, TELEGRAM_WEBHOOK_SECRET, APP_URL",
  );
  process.exit(1);
}

const webhookUrl = `${appUrl}/api/telegram/webhook`;

const res = await fetch(
  `https://api.telegram.org/bot${token}/setWebhook`,
  {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      url: webhookUrl,
      secret_token: secret,
      allowed_updates: ["message", "callback_query"],
    }),
  },
);

const json = await res.json();
if (json.ok) {
  console.log(`✅ Webhook registered: ${webhookUrl}`);
} else {
  console.error("❌ Failed:", json);
  process.exit(1);
}

File: apps/api/scripts/register_webhook.py
"""Run once after deploy to point Telegram to your webhook URL."""
import asyncio
import os
import sys

from dotenv import load_dotenv

load_dotenv()


async def register():
    from telegram import Bot

    token = os.environ["TELEGRAM_BOT_TOKEN"]
    secret = os.environ["TELEGRAM_WEBHOOK_SECRET"]
    app_url = os.environ["APP_URL"]

    bot = Bot(token=token)
    webhook_url = f"{app_url.rstrip('/')}/webhook"

    await bot.set_webhook(
        url=webhook_url,
        secret_token=secret,
        allowed_updates=["message", "callback_query"],
    )
    info = await bot.get_webhook_info()
    print(f"Webhook set: {info.url}")
    print(f"Pending updates: {info.pending_update_count}")


if __name__ == "__main__":
    asyncio.run(register())
